crates/rust-analyzer/src/main_loop/handlers.rs

```diff
@@ -696,14 +694,21 @@ pub fn handle_formatting(
 pub fn handle_code_action(
     world: WorldSnapshot,
     params: lsp_types::CodeActionParams,
-) -> Result<Option<CodeActionResponse>> {
+) -> Result<Option<Vec<lsp_ext::CodeAction>>> {
     let _p = profile("handle_code_action");
+    // We intentionally don't support command-based actions, as those either
+    // requires custom client-code anyway, or requires server-initiated edits.
+    // Server initiated edits break causality, so we avoid those as well.
+    if !world.config.client_caps.code_action_literals {
```

Comment:
cc @kjeremy 

I think, given that pure commands require us to implement client-specific logic anyway, I figured that we might as well just not support them? This is related to the caps PR

Reply:
I think that's only true until we support `workspace/executeCommand`.

Reply:
https://github.com/microsoft/language-server-protocol/issues/724 is the upstream issue. 

>I think that's only true until we support `workspace/executeCommand`.

I personally think that we should not support this :-) The only command that the server can execute is sending workspace edit to the client, but that's strictly worse than just sending it with edit. 

Ie, I think it is confirming to the protocol and a good tradeoff to just not support clients that can do commands with edits.